### ISSUE TITLE
[RFC] Macro Defs - Methods in Macro land

### ISSUE THREAD

**Issue by Blacksmoke16** [crystal-lang/crystal#8835]
As someone who does a good amount with macros, one thing I find lacking is the ability to be DRY.  There currently isn't a way (AFAIK) that allows encapsulating logic to be reused within different macros.  I.e. if there is some piece of logic you need to do multiple times in a macro, you would have to duplicate it.

```cr
{% begin %}
  # Define a temp hash to handle overriding services
  {% services = {} of Nil => Nil %}

  {% for service in ADI::Service.includers %}
    {% raise "#{service.name} includes `ADI::Service` but is not registered.  Did you forget the annotation?" if (annotations = service.annotations(ADI::Register)) && annotations.empty? && !service.abstract? %}
    {% for ann in annotations %}
      {% key = ann[:name] ? ann[:name] : service.name.split("::").last.underscore %}
      {% services[key.id] = {"public" => ann[:public], "type" => service.id} %}
    {% end %}
  {% end %}

  # Define a `getter` in the container for each registered service.
  {% for name, service in services %}
    # Only make the getter public if the service is public
    {% if service["public"] != true %} private {% end %}getter {{name}} : {{service["type"]}}
  {% end %}
{% end %}
```

In this example, I'm defining getters based on the types that include a module, and have an annotation.  I'm also checking for some data within the annotation to determine the name the getter should use, and if it should be public or not. It would be fantastic if you do define a "macro def" to allow reusing the code.

```cr
macro def generate_getters(service_list)
  {% begin %}
    # Define a temp hash to handle overriding services
    {% services = {} of Nil => Nil %}

    {% for service in service_list %}
      {% raise "#{service.name} includes `ADI::Service` but is not registered.  Did you forget the annotation?" if (annotations = service.annotations(ADI::Register)) && annotations.empty? && !service.abstract? %}
      {% for ann in annotations %}
        {% key = ann[:name] ? ann[:name] : service.name.split("::").last.underscore %}
        {% services[key.id] = {"public" => ann[:public], "type" => service.id} %}
      {% end %}
    {% end %}

    # Define a `getter` in the container for each registered service.
    {% for name, service in services %}
      # Only make the getter public if the service is public
      {% if service["public"] != true %} private {% end %}getter {{name}} : {{service["type"]}}
    {% end %}
  {% end %}
end
```

Then the code could be reused, either in the same macro hook, or wherever.

```cr
generate_getters ADI::Service.includers
or
generate_getter [SomeService, FooService]
```

The main use case, for this example at least, would be allowing users to reopen the type the getters are being added to, in order to add custom logic to add additional getters.  It would be much simpler if all they had to do was call `generate_getters` and pass the array of types, versus duplicating that whole code block.

**Comment by asterite** [comment]
This is basically extending the macro language to allow user-defined macro methods. Right now they are hardcoded in the compiler and not extensible. Right?

My idea is that they are also macros. For example, you could reopen `Crystal::StringLiteral` to add methods to it:

```crystal
class Crystal::StringLiteral
  macro foo(x)
    "foo#{x}"
  end
end
```

Then call it:

```crystal
macro bar
  {{ "x".foo }}
end
```

It only makes sense to call the macro method `foo` inside macro code because that returns ASTNode, not code, though maybe we can allow it outside macro code and the code is then transformed to code by calling `to_s` on it.

So the way I'd like to see this implemented is by just allowing to find macros when you call them inside macro code.

That said, I'd like to avoid so many macro code in Crystal, and not having this is a great way to accomplish this.

**Comment by Blacksmoke16** [comment]
> This is basically extending the macro language to allow user-defined macro methods.

Not necessarily.  I would rather just having methods within macro land.  I.e. you pass it one or more `ASTNode`, and it returns an `ASTNode`.

Like your example wouldasterite be

```cr
macro def foo(x : StringLiteral) : StringLiteral
  "foo#{x}"
end

macro bar
  {{ foo "x" }}
end

bar # => "foox"
```

I don't really think it's necessary to be able to reopen the `ASTNodes` to add your own method like you can with the stdlib types.  Using what is currently defined is usually good enough.  All I'm suggesting is being able to have some sort of way to keep things DRY.  As in this latest example, if you had to format `x` three times you would have to do `"foo#{x}"` three times as well.

**Comment by watzon** [comment]
As someone that's also been doing a lot with macros, especially lately, I love this idea. I actually needed something exactly like this last night and was out of luck. I also love @asterite's idea of being able to open up macro definitions and add methods to them like you could with any other class or struct. Would this all be insanely hard to do?

Additionally having initializers for ASTNode types would be nice too. Currently the only way to define a HashLiteral, ArrayLiteral, or anything else afaik is to actually use the literal notation. That means that using a macro to generate a Hash or Array can get kinda hacky.

**Comment by asterite** [comment]
Well, not crazy hard, I managed to implement a quick prototype in an hour.

After thinking a bit more about this, I think it would be really nice to go forward with this.

One thing I dislike about macros is how long and hard to understand they can get. That's a good way to try to avoid them. But if we can Dry them up and make them more concise and easier to understand, and more powerful, maybe it's the best solution.

**Comment by watzon** [comment]
100% agree. Crystal macros are powerful, but there are definitely some instances where you have to do some pretty hacky/unreadable things to do what you need to do. Documentation is also sparse, but once macros are somewhat finalized for v1 @Blacksmoke16 and I can probably work on that.

I feel like having initializers for ASTNode types would go a long way towards cleaning things up, as well as adding some methods that are missing on their standard library counterparts. Also, annotations need some work, but that's another issue.

**Comment by asterite** [comment]
> I feel like having initializers for ASTNode types

Do you have an example of that? I don't quite understand what this means.

**Comment by Blacksmoke16** [comment]
```cr
{% hash = HashLiteral.new %}
versus
{% hash = {} of Nil => Nil %}
```

**Comment by watzon** [comment]
Exactly. Especially if we had stuff like `ArrayLiteral.build` and `HashLiteral.new` which takes a block and allows you to build it.

**Comment by asterite** [comment]
You can already build hashes with "regular" crystal code:

```crystal
enum SomeEnum
  Red
  Green
  Blue
end

{% begin %}
  {% h = {} of Nil => Nil %}
  {% for c in SomeEnum.constants %}
    {% h[c.stringify] = c.stringify.downcase %}
  {% end %}
  p({{ h }})
{% end %}
```

So you see I create a hash and add elements to it at compile-time, instead of outputting a hash literal.

And with macro methods this would be simpler because you could define a method to do that, though you'd need to have `each`.

I don't think we'll add `build` and such methods, it's just too much in my opinion. We can't keep adding built-in macro methods.

**Comment by Blacksmoke16** [comment]
Now that `1.0.0` is release maybe we can revisit this? :slightly_smiling_face:.

**Comment by Blacksmoke16** [comment]
Wanting to bring this RFC up again, as it seems to be something I constantly want/need. Would be open to sponsoring someone to implement this as well :pray:.

To summarize I think the primary goal of this feature is to:

> Reduce the amount of duplication required in macro code by allowing reusable macro methods

For example:

```cr
macro def foo(x)
  "foo#{x}"
end

macro bar
  {{ foo "x" }}
end

bar # => "foox"
```
This plus being able to call the macro def recursively would be a _HUGE_ win to how macro code could be written, ultimately making it easier to read/maintain/share.

Going further and allowing users to monkey patch methods into the stdlib's macro types _could_ be useful, but I don't think it's a requirement, at least for a first pass, given you could just pass the node as an argument. Similarly, supporting parameter/return type restrictions could be a nice win from a documentation/readability POV, but also don't think it should be a blocker to a first pass given normal macros don't support them on parameters either.

Based on what I read in this issue and the related PRs, I think what we'd need to do is like:

* ([ref](https://github.com/crystal-lang/crystal/pull/8836#issuecomment-614075492)) Support parsing `macro def macroName`, storing them in their own place
* ([ref](https://github.com/crystal-lang/crystal/pull/9091#issuecomment-615463064)) Figure out how to handle `@type.foo` given it could clash with methods defined on `TypeNode`
* The actual logic of being able to call the macro defs in existing macros, disallowing them to be used outside of a macro context, but still allowed to use them in different macro contexts. E.g. `SomeNamespace.my_macro_def` to support sharing of the macro code as a shard
* Update doc generation to include a dedicated section for macro defs
* (nice to have) Support `next` in macro contexts
* (nice to have) Support type restrictions on parameters/return
* (future iteration) Possibly allow monkey patching macro defs into existing macro AST types?

Of course not all of this has to be done at once, getting a solid MVP implementation going first would be fine as it would allow people to start using it, then figure out what the pain points are to improve on.

**Comment by HertzDevil** [comment]
Minor: the term "macro def" already refers to [defs that use `@type`](https://crystal-lang.org/reference/1.3/syntax_and_semantics/macros/macro_methods.html).

**Comment by straight-shoota** [comment]
Yeah, we'll have to take care of using clear terms.
We have `def` with macro code, `macro`, top-level macro expressions, and with this feature there's another kind of macro. What should it be called? I'm not sure. `macro def` as keyword would probably be confusing.
From a technical point of view, the `macro` keyword could actually work for this. The difference to normal macros is determined at call site (is it called from normal code or from a macro expression?). But of course it would be helpful to have a clear distinction at the definition.

Similarly, we need to worry about ambiguity at call site, as well. We already have macro calls. How do we call calls in macro expressions?

**Comment by j8r** [comment]
Or call them macro functions - `macro fun`? Technically they are more functions, since they are not methods called from an object instance. `fun` already exist for bindings. In a way the macro-land is already a bit special as there are `for` loops.

**Comment by Blacksmoke16** [comment]
> What should it be called?

Kind of just spitballing here so TBD on exact names of things, but some ideas I thought of:

Use an annotation to denote a macro as "callable". Would require being able to read annotations on macros, and allow displaying annotations applied to them in some way. Would probably require less? compiler work since you could reuse existing macro parsing logic, and just set a flag on it, like `visibility`?

```cr
@[Callable]
macro foo(x)
  "foo#{x}"
end
```

Make the keyword more akin to a modifier of the macro, e.g. `private` or `protected`. Probably more readable and could also just be a flag.

```cr
callable macro foo(x)
  "foo#{x}
end
```

Use some symbol as part of the macro name to denote it being special. Similar could be used to just set a flag, tho I'd say less readable and easier to miss.

```cr
macro $foo(x)
  "foo#{x}
end
```

Require type restrictions on them and use that as a means to differentiate:

```cr
macro foo(x : StringLiteral) : StringLiteral
  "foo#{x}
end
```

Or lastly, use some entirely new non-ambiguous name, like `ast (fun|def) foo(x)` or `node (fun|def) foo(x)` or something like that. Coming up with a clearer work than `macro` might be challenging.

@HertzDevil I'm assuming that logic would be specific to `Call`? I.e. if you tried to do `{{ {{ 2 + 2 }} }}`, you'd still get `Error: can't nest macro expressions`?

**Comment by asterite** [comment]
What's wrong with `macro def`?

**Comment by Blacksmoke16** [comment]
@asterite The concern is mainly around https://github.com/crystal-lang/crystal/issues/8835#issuecomment-1080024082. I.e. that we already have `macro def` meaning something, and making it mean something else would be confusing.

**Comment by straight-shoota** [comment]
Let's settle the terminology in #11945

**Comment by HertzDevil** [comment]
> @HertzDevil I'm assuming that logic would be specific to `Call`? I.e. if you tried to do `{{ {{ 2 + 2 }} }}`, you'd still get `Error: can't nest macro expressions`?

It is as shown above:

* If the macro is not already within a macro evaluation context, then `2.+(2)` is not a valid `macro def` call, because the first `2` is not a type.
* If it is within a macro evaluation context, possibly yet another outer `{{ ... }}`, the error would be something like `expected Call, got MacroExpression`.

**Comment by HertzDevil** [comment]
Also minor point: the following is valid syntax:

```crystal
module A
  macro def
    1
  end
end

A.def # => 1
```

**Comment by asterite** [comment]
I don't understand what's the ambiguity.

**Comment by HertzDevil** [comment]
Here are some additional examples involving non-printing or escaped macros:

```crystal
macro def foo
  {% x = 1 %} # Error: macro def call must use {{ ... }}, not {% ... %}
end

macro foo
  {{ {% foo %} }}     # Error: macro def call must use {{ ... }}, not {% ... %}
  {% x = {{ foo }} %} # okay
end
```

`{% ... %}` cannot be nested within `{{ ... }}`. The opposite is allowed.

Also notice how the macro and the `macro def` are both called `foo`. This is technically fine, because the nested macro treatment implies they are always in separate namespaces, but we may want to forbid this to avoid confusion. This is notably _not_ forbidden between defs and macros already; macros are "ordered before" defs in terms of "overload" ordering if they share the same name.

```crystal
macro foo
  \{{ {{ 1 + 2 }} }}
  {% debug %}
end

foo # => {{ 3 }}
```

If an outer `{{ ... }}` is escaped, everything within is interpolated normally. This is existing behavior and will continue to be allowed. The body of a `macro def` is never considered to be escaped in this manner.

```crystal
macro def foo
  \{{ 1 + 2 }} # Error: unknown token: '{'
end

macro bar
  {% x = \{{ 1 + 2 }} %} # Error: unknown token: '{'
end

bar
```

Escaped macros cannot be nested in macros, so it follows bare escaped ones are also disallowed within `macro def`s. At the moment the supported AST expressions within a `macro def` should be exactly what are currently permitted within `{% ... %}` (no `case` for example).

***

> I don't understand what's the ambiguity.

[This one](https://github.com/crystal-lang/crystal/pull/9091#issuecomment-615463064), i.e. whether `macro def`s and AST node methods should both be callable with the same syntax.

**Comment by HertzDevil** [comment]
It still clashes with the names of AST node methods:

```crystal
struct Int32
  macro def ancestors
  end
end

{% Int32.ancestors %} # does this return a `Nop`?
```

**Comment by lbguilherme** [comment]
We can either accept that these methods can be overridden (just like regular methods from stdlib can), or perhaps introduce some less aggressive way to avoid the ambiguity: adding a `call` macro:

```cr
struct Int32
  macro def foo(a, b)
    a + b
  end
end

{% Int32.call(:foo, 1, 2) %} # returns a `3` literal
```

This is easier to read and integrates better with existing syntax, IMHO. It would be defined for TypeNodes and at the top level.

**Comment by asterite** [comment]
> It still clashes with the names of AST node methods

Right, and in that case you are overriding the method. There's still no ambiguity.

> The opposite argument is we can then redefine all AST node methods as "primitive" macro defs, but that's a much larger undertaking and I'm not sure if a breaking change is avoidable there

I don't think there's a breaking change at all.

All the current macro methods are defined with `macro def` and they are primitives. Of course that syntax doesn't exist yet, so you can't do it, but we can imagine that's how it's done.

Now we are enabling `macro def` to be actually defined by users, and the existing ones can also be defined, it's just that they are `@[Primitive]` somehow (and eventually we could implement them like that)

So with this:

```crystal
struct Int32
  macro def ancestors
  end
end
```

you are overriding the behavior of `Int32.ancestors` inside macros, but `String.ancestors` would still return the "correct" value.

Then we can redefine ancestors for every type like this:

```crystal
class Crystal::Macros::TypeNode
  macro def ancestors
  end
end
```

It's a bit similar to how we can define a method for all classes by defining an instance method in `Class`:

```crystal
class Class
  def foo
    1
  end
end

p Int32.foo # It works!
```

So there's a bit of "conflict" in that you can define a macro def on a type, or on TypeNode, and both kind of work, and it's exactly the same as in regular Crystal code, where you can define a class method on a type, or a method on Class, and they both kind of work.

Then you can define macro methods on `Crystal::Macros::*`, so for example adding a method to a `StringLiteral`. And all existing methods are also `macro def`, they are just primtiives.

Is there any conflict or ambiguity with that approach?

**Comment by HertzDevil** [comment]
It seems the idea here is to establish a "metaclass" hierarchy on top of `TypeNode` that mirrors the real hierarchy of metaclasses. I am still not sure if that's the best approach, but it does admit an alternative syntax:

```crystal
macro class TypeNode
  @[Primitive]
  def ancestors
  end
end

macro class Int32
  # the entire body deals with the `Int32` macro "metaclass", so
  # every def is implicitly a `macro def`

  def ancestors
  end
end

# not allowed for macros, but allowed for the non-macro class analog (#11764)
macro def String.ancestors
end
```

And we can simply call them "defs" of "macro metaclasses", avoiding the terminology conflict altogether.

**Comment by asterite** [comment]
Oh, I like that! I like that these things are "scoped" inside a "macro" context.

But how would you define a top-level macro method with that syntax? Is it just `macro def`? I'm thinking I'd still like methods inside `macro class` to be called `macro def` just because if you see `def ...` without the surrounding context you can't immediately know if that's a method to be used at compile-time or not.

So maybe we only need `macro def` in that case, and `macro class` isn't really needed, because the type `Int32` is the same inside or outside macros.

**Comment by Blacksmoke16** [comment]
Would `macro def` also be used to define them within a non stdlib namespace as well? Or is the idea you'd also be able to do like `macro class MyModule`? But yea, if we're okay with just allowing the end user to override macro methods like they could with stdlib methods, then that syntax would definitely be more readable/flexible I'd say. I'm all for having an implementation that is a bit future facing to make the macro language implementation simpler longer term, which sounds like this approach would.

**Comment by HertzDevil** [comment]
A slightly different idea: we reserve `macro class` to represent the AST node types in the macro language, without the `Crystal::Macros` part. After all, the types we use in the macro `#is_a?` are in an entirely separate namespace, do not physically reside under `Crystal::Macros`, and never conflict with non-macro types sharing the same names. Then normal types use `macro def` within the usual `class`. Example:

```crystal
macro class ArrayLiteral
  @[Primitive(...)]
  macro def each_with_index(&); end

  @[Primitive(...)]
  macro def size; end

  macro def empty?
    # note: implicit `self`
    size == 0
  end

  macro def splat(trailing_string = nil)
    str = ""
    each_with_index do |v, i|
      str = "#{str.id}, " if i > 0
      str = "#{str.id}#{v}"
    end
    if trailing_string && !empty?
      str = "#{str.id}#{trailing_string.id}"
    end
    str.id
  end
end

# okay, not the AST node macro type
class ArrayLiteral
  macro def empty?
    # same as `ArrayLiteral.ancestors.empty?` in a macro context
    ancestors.empty? # => false
  end
end

# okay, not the AST node macro type either
class Crystal::Macros::ArrayLiteral
end
```

**Comment by asterite** [comment]
I like that! In the API docs I can imagine they will also live inside a different section.

**Comment by lbguilherme** [comment]
Inside a `macro class` you define `macro def`s that happen to be like instance methods on the AST node. At the same time `macro def`s at a regular class/module operate like a metaclass method. This is odd because the same syntax create methods that are called in different ways. I propose we use regular method syntax inside a `macro class` to define methods of AST nodes (as primitives), and `macro def` syntax at regular classes/modules to define methods that are callable from the TypeNode.

```cr
macro class ArrayLiteral
  def foo
    "foo"
  end
end

class String
  macro def bar
    ancestors.foo
  end
end

{% p String.bar %} # prints "foo"
```

**Comment by Blacksmoke16** [comment]
Anyone interested in getting a PoC together? Would love to see this gain some momentum.

**Comment by HertzDevil** [comment]
The other day I realized the current macro defs are named like that because methods referring to `{{ @type }}` used to have the `macro def` syntax: #2565

**Comment by Blacksmoke16** [comment]
@HertzDevil Yea, that's why I was thinking we could reuse that syntax since nothing is using it anymore. Tho, as you and others pointed out, there are other options depending on exact path we want to take this. I.e. The more focused approach of "just allow macro code to be reused by defining reusable macro methods" or treat this feature as a way to implement the existing macro methods/modify the methods available on the macro AST nodes.

**Comment by stellarpower** [comment]
May be a stupid question, but, would it be physically possible to force a macro to expand within the context of another macro? A bit like how I can call `.id` on a macro's argument to . Would it be an option to tell the compiler forcibly to assume that a token is a macro, a constant, etc. and to expand it out and then proceed who expand the current macro with that value treated the same way it would if passed a literal? e.g. `Crystal::Macros::ASTNode#eval` or a different liquid-like syntax (IDK, `{! !}`) that would demarcate the contained expression as something to be expanded into the current macro scope.
